File: packages/core/src/paths.ts
/**
 * Cross-platform path resolution (shared between CLI and core)
 *
 * Single source of truth for all file paths. CLI primes at startup
 * with overrides; core enrichment code calls paths() to get cached result.
 *
 * Precedence (highest to lowest):
 *  1. CLI flags (PathOverrides)
 *  2. Category env vars (LET_DATA_DIR, LET_CONFIG_DIR, etc.)
 *  3. LET_HOME or TOOLS_HOME env var, defaulting to ~/.tools/let
 */

import { homedir } from 'node:os';
import { isAbsolute, join } from 'node:path';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface PathOverrides {
	dataDir?: string | undefined;
	configDir?: string | undefined;
	cacheDir?: string | undefined;
	sourcesDir?: string | undefined;
}

export interface ResolvedPaths {
	/** Directory containing config file */
	config: string;
	/** Directory containing let.db */
	data: string;
	/** Directory containing {portalId}/ cache entries */
	cache: string;
	/** Directory containing *.db source databases */
	sources: string;
}

export interface DerivedPaths {
	/** Config file path (let.config.toml) */
	configFile: string;
	/** Template config file path (dev only) */
	templateFile: string;
	/** .env file path */
	envFile: string;
	/** Listings database path */
	database: string;
	/** Listings database backup path */
	backup: string;
	/** JSON export path */
	jsonExport: string;
	/** Source database path for a given name */
	sourceDb(name: string): string;
	/** Cache directory for a portal ID */
	cacheDir(id: string): string;
	/** Cache entry (data.json) for a portal ID */
	cacheEntry(id: string): string;
}

// ---------------------------------------------------------------------------
// Resolution
// ---------------------------------------------------------------------------

function resolveAbsoluteOrCwd(value: string): string {
	return isAbsolute(value) ? value : join(process.cwd(), value);
}

/**
 * Build derived paths from resolved directories.
 */
function buildDerived(resolved: ResolvedPaths): DerivedPaths {
	const configFileName = 'let.config.toml';
	return {
		configFile: join(resolved.config, configFileName),
		templateFile: join(resolved.config, 'let.config.template.toml'),
		envFile: join(resolved.config, '.env'),
		database: join(resolved.data, 'let.db'),
		backup: join(resolved.data, 'let.db.bak'),
		jsonExport: join(resolved.data, 'let.db.json'),
		sourceDb(name: string): string {
			return join(resolved.sources, `${name}.db`);
		},
		cacheDir(id: string): string {
			return join(resolved.cache, id);
		},
		cacheEntry(id: string): string {
			return join(resolved.cache, id, 'data.json');
		},
	};
}

// ---------------------------------------------------------------------------
// Cached singleton
// ---------------------------------------------------------------------------

let cached: { resolved: ResolvedPaths; derived: DerivedPaths } | null = null;

type DirSet = { config: string; data: string; cache: string; sources: string };

/**
 * Compute base defaults from LET_HOME / TOOLS_HOME, defaulting to ~/.tools/let.
 */
function resolveDefaults(): { defaults: DirSet } {
	const letHome = process.env['LET_HOME'] || join(process.env['TOOLS_HOME'] || join(homedir(), '.tools'), 'let');
	return {
		defaults: {
			config: join(letHome, 'data'),
			data: join(letHome, 'data'),
			cache: join(letHome, 'cache'),
			sources: join(letHome, 'sources'),
		},
	};
}

/**
 * Apply category env var if set, otherwise return the default value.
 */
function envOrDefault(envKey: string, fallback: string): string {
	const value = process.env[envKey];
	return value ? resolveAbsoluteOrCwd(value) : fallback;
}

/**
 * Resolve all paths using the precedence chain.
 * Result is cached after first call. Subsequent calls return the cache
 * unless overrides are provided (which forces re-resolution).
 */
export function resolvePaths(overrides?: PathOverrides): { resolved: ResolvedPaths; derived: DerivedPaths } {
	if (cached && !overrides) return cached;

	const { defaults } = resolveDefaults();

	// Apply category env vars over defaults
	const resolved: ResolvedPaths = {
		config: envOrDefault('LET_CONFIG_DIR', defaults.config),
		data: envOrDefault('LET_DATA_DIR', defaults.data),
		cache: envOrDefault('LET_CACHE_DIR', defaults.cache),
		sources: envOrDefault('LET_SOURCES_DIR', defaults.sources),
	};

	// Apply CLI flag overrides (highest priority)
	if (overrides?.configDir) resolved.config = resolveAbsoluteOrCwd(overrides.configDir);
	if (overrides?.dataDir) resolved.data = resolveAbsoluteOrCwd(overrides.dataDir);
	if (overrides?.cacheDir) resolved.cache = resolveAbsoluteOrCwd(overrides.cacheDir);
	if (overrides?.sourcesDir) resolved.sources = resolveAbsoluteOrCwd(overrides.sourcesDir);

	const result = { resolved, derived: buildDerived(resolved) };
	cached = result;
	return result;
}

/**
 * Get cached resolved paths. Throws if resolvePaths() has not been called yet.
 */
export function paths(): { resolved: ResolvedPaths; derived: DerivedPaths } {
	if (!cached) {
		// Auto-resolve with no overrides (e.g. core enrichment calling before CLI priming)
		return resolvePaths();
	}
	return cached;
}

/**
 * Reset the cached singleton (test-only).
 */
export function resetPaths(): void {
	cached = null;
}

File: crates/let-sdk/src/paths.rs
#![forbid(unsafe_code)]

use std::path::{Path, PathBuf};

#[derive(Debug, Clone, Default)]
pub struct PathOverrides {
    pub data_dir: Option<PathBuf>,
    pub config_dir: Option<PathBuf>,
    pub cache_dir: Option<PathBuf>,
    pub sources_dir: Option<PathBuf>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ResolvedPaths {
    pub config: PathBuf,
    pub data: PathBuf,
    pub cache: PathBuf,
    pub sources: PathBuf,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DerivedPaths {
    pub config_file: PathBuf,
    pub template_file: PathBuf,
    pub env_file: PathBuf,
    pub database: PathBuf,
    pub backup: PathBuf,
    pub json_export: PathBuf,
}

impl DerivedPaths {
    pub fn source_db(&self, sources_root: &Path, name: &str) -> PathBuf {
        sources_root.join(format!("{name}.db"))
    }

    pub fn cache_dir(&self, cache_root: &Path, id: &str) -> PathBuf {
        cache_root.join(id)
    }

    pub fn cache_entry(&self, cache_root: &Path, id: &str) -> PathBuf {
        cache_root.join(id).join("data.json")
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PathBundle {
    pub resolved: ResolvedPaths,
    pub derived: DerivedPaths,
}

fn make_absolute(path: PathBuf) -> PathBuf {
    if path.is_absolute() {
        path
    } else {
        std::env::current_dir()
            .unwrap_or_else(|_| PathBuf::from("."))
            .join(path)
    }
}

fn env_or_default(key: &str, fallback: PathBuf) -> PathBuf {
    std::env::var_os(key)
        .map(PathBuf::from)
        .map(make_absolute)
        .unwrap_or(fallback)
}

fn default_home() -> PathBuf {
    if let Some(let_home) = std::env::var_os("LET_HOME") {
        return PathBuf::from(let_home);
    }

    if let Some(tools_home) = std::env::var_os("TOOLS_HOME") {
        return PathBuf::from(tools_home).join("let");
    }

    dirs::home_dir()
        .unwrap_or_else(|| PathBuf::from("~"))
        .join(".tools")
        .join("let")
}

fn build_derived(resolved: &ResolvedPaths) -> DerivedPaths {
    DerivedPaths {
        config_file: resolved.config.join("let.config.toml"),
        template_file: resolved.config.join("let.config.template.toml"),
        env_file: resolved.config.join(".env"),
        database: resolved.data.join("let.db"),
        backup: resolved.data.join("let.db.bak"),
        json_export: resolved.data.join("let.db.json"),
    }
}

pub fn resolve_paths(overrides: Option<PathOverrides>) -> PathBundle {
    let root = default_home();
    let defaults = ResolvedPaths {
        config: root.join("data"),
        data: root.join("data"),
        cache: root.join("cache"),
        sources: root.join("sources"),
    };

    let mut resolved = ResolvedPaths {
        config: env_or_default("LET_CONFIG_DIR", defaults.config),
        data: env_or_default("LET_DATA_DIR", defaults.data),
        cache: env_or_default("LET_CACHE_DIR", defaults.cache),
        sources: env_or_default("LET_SOURCES_DIR", defaults.sources),
    };

    if let Some(ovr) = overrides {
        if let Some(dir) = ovr.config_dir {
            resolved.config = make_absolute(dir);
        }
        if let Some(dir) = ovr.data_dir {
            resolved.data = make_absolute(dir);
        }
        if let Some(dir) = ovr.cache_dir {
            resolved.cache = make_absolute(dir);
        }
        if let Some(dir) = ovr.sources_dir {
            resolved.sources = make_absolute(dir);
        }
    }

    PathBundle {
        derived: build_derived(&resolved),
        resolved,
    }
}

pub fn paths() -> PathBundle {
    resolve_paths(None)
}

#[cfg(test)]
mod tests {
    use super::{PathOverrides, resolve_paths};
    use std::path::PathBuf;

    #[test]
    fn cli_override_wins() {
        let bundle = resolve_paths(Some(PathOverrides {
            data_dir: Some(PathBuf::from("/tmp/let-data")),
            ..PathOverrides::default()
        }));
        assert_eq!(bundle.resolved.data, PathBuf::from("/tmp/let-data"));
    }

    #[test]
    fn derived_paths_are_consistent() {
        let bundle = resolve_paths(Some(PathOverrides {
            data_dir: Some(PathBuf::from("/tmp/let-data")),
            config_dir: Some(PathBuf::from("/tmp/let-config")),
            ..PathOverrides::default()
        }));

        assert_eq!(
            bundle.derived.database,
            PathBuf::from("/tmp/let-data/let.db")
        );
        assert_eq!(
            bundle.derived.config_file,
            PathBuf::from("/tmp/let-config/let.config.toml")
        );
    }

    #[test]
    fn consecutive_override_calls_do_not_reuse_previous_values() {
        let first = resolve_paths(Some(PathOverrides {
            data_dir: Some(PathBuf::from("/tmp/let-data-a")),
            ..PathOverrides::default()
        }));
        let second = resolve_paths(Some(PathOverrides {
            data_dir: Some(PathBuf::from("/tmp/let-data-b")),
            ..PathOverrides::default()
        }));

        assert_eq!(first.resolved.data, PathBuf::from("/tmp/let-data-a"));
        assert_eq!(second.resolved.data, PathBuf::from("/tmp/let-data-b"));
    }
}
